internal/lifecycle/readme.go

```diff
@@ -0,0 +1,127 @@
+
+package lifecycle
+
+import (
+	"bufio"
+	"fmt"
+	"github.com/GoogleCloudPlatform/serverless-sample-tester/internal/util"
+	"os"
+	"os/exec"
+	"regexp"
+	"strings"
+)
+
+// The tag that should appear immediately before code blocks in a README to indicate that the enclosed commands
+// are to be used by this program for building and deploying the sample.
+var codeTag = "sst-run-linuxmacos"
+
+// parseREADME parses a README file with the given name. It reads terminal commands surrounded by one of the codeTags
+// listed above and loads them into a Lifecycle. In the process, it replaces the Cloud Run service name and Container
+// Registry tag with the provided inputs.
+func parseREADME(filename, serviceName, gcrURL string) (Lifecycle, error) {
+	file, err := os.Open(filename)
+	if err != nil {
+		return nil, err
+	}
+	defer file.Close()
+
+	scanner := bufio.NewScanner(file)
+
+	var lifecycle Lifecycle
+
+	inCode := false
+	for scanner.Scan() {
+		line := strings.TrimSpace(scanner.Text())
+
+		if strings.Contains(line, "```") && inCode {
+			inCode = false
+		} else if strings.Contains(line, codeTag) {
+			scanner.Scan()
+			startCodeBlockLine := scanner.Text()
+
+			if !strings.Contains(startCodeBlockLine, "```") {
+				return nil, fmt.Errorf("[lifecycle.parseREADME] parsing README: incorrect format")
+			}
+
+			inCode = true
+		} else if inCode {
+			for line[len(line)-1] == '\\' {
+				line = line[:len(line)-1]
+
+				scanner.Scan()
+				line = line + strings.TrimSpace(scanner.Text())
+			}
+
+			line = os.ExpandEnv(line)
+			line = replaceGCRURL(line, gcrURL)
+			line = replaceServiceName(line, serviceName)
+			sp := strings.Split(line, " ")
+
+			var cmd *exec.Cmd
+			if strings.Contains(line, "gcloud") {
+				a := append(util.GcloudCommonFlags, sp[1:]...)
+				cmd = exec.Command("gcloud", a...)
+			} else {
+				cmd = exec.Command(sp[0], sp[1:]...)
+			}
+
+			lifecycle = append(lifecycle, cmd)
+		}
+	}
+
+	if inCode {
+		return nil, fmt.Errorf("[lifecycle.parseREADME] parsing README: incorrect format")
+	}
+
+	if err := scanner.Err(); err != nil {
+		return nil, fmt.Errorf("[lifecycle.parseREADME] README bufio.Scanner: %w", err)
+	}
+
+	if len(lifecycle) == 0 {
+		return nil, fmt.Errorf("[lifecycle.parseREADME] parsing README: no commands found")
+	}
+
+	return lifecycle, nil
+}
+
+// replaceServiceName takes a terminal command string as input and replaces the Cloud Run service name, if any.
+// It detects whether the command is a gcloud run command and replaces the last argument that isn't a flag
+// with the input service name.
+func replaceServiceName(command, serviceName string) string {
+	containsGcloud, _ := regexp.MatchString(`\bgcloud\b`, command)
+	containsRun, _ := regexp.MatchString(`\brun\b`, command)
+
+	if !(containsGcloud && containsRun) {
+		return command
+	}
+
+	sp := strings.Split(command, " ")
+	for i := len(sp) - 1; i >= 0; i-- {
+		if !strings.Contains(sp[i], "--") {
+			sp[i] = serviceName
```

Comment:
Can gcloud flags have args or are they always in the `flag=val` format?

```
gcloud run --flag1 arg1 serviceName --flag2 arg2
```

Reply:
I believe they can have args and can be in the format followed in your example.

@danielylee01 plans to revise the Cloud Run service name detection and replacing algorithm in a future PR.

Reply:
SGTM - let's open an issue then for this PR to make sure we don't loose track. We can close it with a future PR once complete.

Reply:
Okay, planning to open an issue entitled "Cloud Run service name detection and replacing algorithm flawed" after this current PR gets merged.